### DISCUSSION TITLE
Cant open Parker connection

### DISCUSSION THREAD

**Discussion by flynnd3** [pymeasure/pymeasure#620] [Q&A]
I'm sorry for asking this here as it's probably not the place for it, but can someone give me some example code to connect to a parker drive as I'm having no luck.
I'm new to programming and would place myself in the low end of the intermediate level, but have a need to connect to a Parker GV6 drive.
running python 3.9 on Windows 10 and have had trouble with this.

I have tried multiple variations to get a connection to the drive, and the one that got the furthest with the following:

`from pymeasure.instruments.parker import Parker.GV6 as drive`
`test = drive("COM2")`

after that last command, I get the following.

`Traceback (most recent call last):`
 `File "<stdin>", line 1, in <module>`
 `File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 40, in __init__`
  `  super(ParkerGV6, self).__init__(`
 `File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\instrument.py", line 69, in __init__`
  `  self.status = self.measurement("*STB?",`
`AttributeError: can't set attribute`

Any help greatly appreciated on this

**Comment by bilderbuchi** [comment]
This is exactly the right place!

Is this the whole error message? It seems incomplete at the end.

It seems you're not using the latest pymeasure version (0.10), could you try upgrading and try again before we proceed, please?

**Reply by flynnd3** [reply]
Hi bilderbuchi,

Thanks for responding and assisting me with this.

first, that was the complete error message. I agree that it did look to be cut off at the end.

I have upgraded to version 0.10 and using my same commands (above), I now get the following Error:

`Traceback (most recent call last):`
`  File "<stdin>", line 1, in <module>`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 44, in __init__
    self.setDefaults()`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\instrument.py", line 209, in __getattribute__  `
`    return super().__getattribute__(name)`
`AttributeError: 'ParkerGV6' object has no attribute 'setDefaults'`

**Reply by bilderbuchi** [reply]
OK, this is more helpful. AFAICT, this is a very old instrument from the beginning of pymeasure. Best try opening the `parkerGV6.py` file in your site-packages, and change [this line](https://github.com/pymeasure/pymeasure/blob/0400c37c9fa474afe8fbddb041de6f5bb708a259/pymeasure/instruments/parker/parkerGV6.py#L44) and [that line](https://github.com/pymeasure/pymeasure/blob/0400c37c9fa474afe8fbddb041de6f5bb708a259/pymeasure/instruments/parker/parkerGV6.py#L75) to use the[ correct method name](https://github.com/pymeasure/pymeasure/blob/0400c37c9fa474afe8fbddb041de6f5bb708a259/pymeasure/instruments/parker/parkerGV6.py#L58) `set_defaults`.
I suspect a renaming did not change all call sites at some point in the past. If that fixes the problem, we would appreciate a  pull request (if you can) to fix this for good.

**Reply by flynnd3** [reply]
this is the latest error.
`Traceback (most recent call last):`
`  File "<stdin>", line 1, in <module>`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 40, in __init__`
`    super().__init__(`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\instrument.py", line 69, in __init__`
`    # Property return itself when invoked from a class`
`AttributeError: can't set attribute`

I shall try this again tomorrow on another PC just to check, and get back to you.

Thanks for all you help

**Reply by bilderbuchi** [reply]
Ha, [this line](https://github.com/pymeasure/pymeasure/blob/0400c37c9fa474afe8fbddb041de6f5bb708a259/pymeasure/instruments/parker/parkerGV6.py#L62) is wrong, `echo` is a method, not a property. This should read `self.echo(False)`.

**Reply by flynnd3** [reply]
Hi bilderbuchi,

here is the error after the most recent change:

`File "<stdin>", line 1, in <module>`
 ` File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 44, in __init__`
`    self.set_defaults()`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 62, in set_defaults`
`    self.echo(False)`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 208, in echo`
`    self.write("ECHO0")`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 50, in write
    self.connection.write(command + "\r")`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\instrument.py", line 209, in __getattribute__`
`    return super().__getattribute__(name)`
`AttributeError: 'ParkerGV6' object has no attribute 'connection'`

**Reply by bilderbuchi** [reply]
@msmttchr any idea what is going wrong here? Is the getattribute from the dynamic properties going haywire?

**Reply by msmttchr** [reply]
This should be:
`self.adapter.connection.write(...`
Also I doubt that `read` method is also buggy, I think `self.readlines` should be replaced by `self.adapter.connection.readlines`

**Reply by flynnd3** [reply]
so i have change lines 50 and 56 inparkerGV6.py as per msmttchr's recommendations, and am getting the following.

`>>> test = drive("COM5")`
`Traceback (most recent call last):`
`  File "<stdin>", line 1, in <module>`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 41, in __init__`
`    SerialAdapter(port, 9600, timeout=5),`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\adapters\serial.py", line 51, in __init__`
`    self.connection = serial.Serial(port, **kwargs)`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\serial\serialwin32.py", line 33, in __init__`
`    super(Serial, self).__init__(*args, **kwargs)`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\serial\serialutil.py", line 244, in __init__`
`    self.open()`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\serial\serialwin32.py", line 64, in open`
`    raise SerialException("could not open port {!r}: {!r}".format(self.portstr, ctypes.WinError()))`
`serial.serialutil.SerialException: could not open port 'COM5': PermissionError(13, 'Access is denied.', None, 5)`

**Reply by bilderbuchi** [reply]
That typically happens when some other process/object still has a connection on that port. Is there an old/previous python process still running? Does restarting your terminal al/IDE help?

**Reply by flynnd3** [reply]
restarted the terminal and ran the commands again. Now getting the following:

`>>> test = drive("COM5")`
`Traceback (most recent call last):`
 ` File "<stdin>", line 1, in <module>`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 44, in __init__`
 `   self.set_defaults()`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 62, in set_defaults`
`    self.echo(False)`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 208, in echo`
`    self.write("ECHO0")`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\pymeasure\instruments\parker\parkerGV6.py", line 50, in write
    self.adapter.connection.write(command + "\r")`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\serial\serialwin32.py", line 310, in write`
   ` data = to_bytes(data)`
`  File "C:\Users\flynnd3\AppData\Roaming\Python\Python39\site-packages\serial\serialutil.py", line 65, in to_bytes`
`    raise TypeError('unicode strings are not supported, please encode to bytes: {!r}'.format(seq))`
`TypeError: unicode strings are not supported, please encode to bytes: 'ECHO0\r'`

**Reply by bilderbuchi** [reply]
> self.adapter.connection.write(command + "\r")

This should be `super().write(command + "\r")` (or, failing that, `self.adapter.write(command + "\r")` as pointed out in https://github.com/pymeasure/pymeasure/discussions/620#discussioncomment-2732260. Otherwise, the [encoding step](https://github.com/pymeasure/pymeasure/blob/0400c37c9fa474afe8fbddb041de6f5bb708a259/pymeasure/adapters/serial.py#L58) in SerialAdapter.write is skipped.

**Reply by bilderbuchi** [reply]
At this point it's clear to me that we need to get this instrument fixed, so I've created issue #623. Let us continue there, please.

**Reply by flynnd3** [reply]
Hi bilderbuchi.
The latest edits proposed by msmttchr were not throwing up any errors, though I was not in a position to test further yesterday, I shall be able to report back more info on Monday.
Thanks for all your help so far

**Reply by bilderbuchi** [reply]
Sure! Are you in a position to [submit a pull request](https://pymeasure.readthedocs.io/en/latest/dev/contribute.html) with the required changes so we can fix #623? That would be helpful in unloading us maintainers.

**Reply by flynnd3** [reply]
Not sure how to do a pull request, but I'll look into it Monday and get it done for you.
I'll touch base here if I have any further issues when testing properly 
Thanks

**Reply by bilderbuchi** [reply]
Much appreciated!

> I'll touch base here if I have any further issues when testing properly

Let's continue over in #623, please, to keep further discussion in one place.

**Comment by msmttchr** [comment]
I am sorry, I gave you incorrect suggestion, let me have a second chance:
`self.connection.write(...`
should be:
`self.adapter.write(...`
Also I doubt that `read` method is also buggy, I think line:
https://github.com/pymeasure/pymeasure/blob/0400c37c9fa474afe8fbddb041de6f5bb708a259/pymeasure/instruments/parker/parkerGV6.py#L56

should be replaced as follows:
```
        readlines = b"\n".join(self.connection.readlines()).decode()
        return re.sub(r'\r\n\n(>|\?)? ', '', readlines)
```
I am sorry, I have no chance to try it, but it should be correct now
